[RPC profiling] Allow disableProfiler() to be called from another thread.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44967 Fallback to CPU when remote end does not have CUDA for profiling
* #44923 [RPC profiling] Add tests to ensure RPC profiling works on single threaded
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.

This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

rohan-varma committed:
[RPC profiling] Allow disableProfiler() to be called from another thread. 

This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Allow disableProfiler() to be called from another thread. "


This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]

mrshenli commented on c10/util/ThreadLocalDebugInfo.cpp:
```diff
@@ -51,6 +51,15 @@ std::shared_ptr<DebugInfoBase> ThreadLocalDebugInfo::_pop(DebugInfoKind kind) {
+  TORCH_CHECK(
+    debug_info && debug_info->kind_ == kind,
+    "Expected debug info of type ", (size_t) kind);
+    auto res = debug_info;
```
wrong indent?

mrshenli commented on c10/util/ThreadLocalDebugInfo.cpp:
```diff
@@ -51,6 +51,15 @@ std::shared_ptr<DebugInfoBase> ThreadLocalDebugInfo::_pop(DebugInfoKind kind) {
+    debug_info && debug_info->kind_ == kind,
+    "Expected debug info of type ", (size_t) kind);
+    auto res = debug_info;
+    return res->info_;
```
BTW, why creating a copy of `debug_info` instead of directly return `debug_info->info_`?

rohan-varma replied:
Was just following the above code, but it can just be `debug_info->info_` indeed.

mrshenli commented on c10/util/ThreadLocalDebugInfo.cpp:
```diff
@@ -51,6 +51,15 @@ std::shared_ptr<DebugInfoBase> ThreadLocalDebugInfo::_pop(DebugInfoKind kind) {
 }
 
+/* static */
+std::shared_ptr<DebugInfoBase> ThreadLocalDebugInfo::_peek(DebugInfoKind kind) {
```
Looks like the kind can be a const reference? Please ignore if it is just an integer.

rohan-varma replied:
It looks like it's `uint8_t` so this should be okay

mrshenli commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -468,21 +467,29 @@ void enableProfiler(const ProfilerConfig& new_config) {
-  auto state = c10::ThreadLocalDebugInfo::_pop(c10::DebugInfoKind::PROFILER_STATE);
+  std::shared_ptr<c10::DebugInfoBase> state;
+  if (cleanupTLSState) {
+    state = c10::ThreadLocalDebugInfo::_pop(c10::DebugInfoKind::PROFILER_STATE);
```
nit question: do we have to pop here? Will it work if this line always uses peek, and let the `if (cleanupTLSState)` decide whether it should pop, to save one branch?

rohan-varma replied:
In the majority use case where `cleanupTLSState = true`, that would incur both a pop and a peek, whereas the current would have an extra branch but only do one of pop/peek, right?

mrshenli reviewed (approved):
LGTM!

rohan-varma committed:
Update on "[RPC profiling] Allow disableProfiler() to be called from another thread. "


This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Allow disableProfiler() to be called from another thread. "

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.

This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Allow disableProfiler() to be called from another thread. "

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.

This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Allow disableProfiler() to be called from another thread. "

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.

This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]

ilia-cher commented on torch/csrc/autograd/init.cpp:
```diff
@@ -61,7 +61,11 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
+  m.def(
+      "_disable_profiler",
+      disableProfiler,
+      py::arg("cleanup_tls_states") = true,
```
are there cases when this argument should be false? i thought that we'd want to clean up local pointer to profiler state anyways

rohan-varma replied:
From offline discussion:
1) We set this to `false` in child thread since we don't explicitly push those profiler TLS states, so we shouldn't want to pop them in disableProfiler
2) However, there's no issues if we set to `true`, due to the way `ThreadLocalStateGuard` works.

rohan-varma committed:
Update on "[RPC profiling] Allow disableProfiler() to be called from another thread. "

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.

This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Allow disableProfiler() to be called from another thread. "

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.

This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Allow disableProfiler() to be called from another thread. "

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* #44655 [RPC profiling] Don't wrap toHere() calls with profiling
* **#44653 [RPC profiling] Allow disableProfiler() to be called from another thread.**
* #44646 Remove thread_local RecordFunctionGuard from profiler.

This changes the profiler per a discussion with @ilia-cher offline that enables `disableProfiler()` event consolidation logic to be called from different threads (i.e. threads where the profiler was not explicitly enabled). This is needed to support the functionality enabled by D23638387 where we defer profiling event collection until executing an async callback that can execute on a different thread, to support RPC async function profiling.

This is done by introducing 2 flags `cleanupTLSState` and `consolidate` which controls whether we should clean up thread local settings (we don't do this when calling `disableProfiler()` on non-main threads) and whether we should consolidate all profiled events. Backwards compatiblity is ensured since both options are true by default.

Added a test in `test_misc.cpp` to test this.

Differential Revision: [D23638499](https://our.internmc.facebook.com/intern/diff/D23638499/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23638499/)!

[ghstack-poisoned]
